Large set of changes (see details)

tjaffri opened the pull request:
There are a few large fixes in here. See below:

1. Resolves #50 - Node package naming conventions. All the package.json files are now in this format.
2. Resolves #44 - Thermostat schema in OCF format after many discussions across @jeanpa, @arjun-msft and @jasongin. Thanks everyone for your help!
3. Resolves #46 - Basic OCF schemas. These are checked in under the oic.r.* namespaces, and are referenced from the Thermostat schema.
4. Partial fixes for #45 - Thermostat translator in OCF format. The Get and Post are currently stubbed out, but I need to fix #52 before I implement these. Wanted to get this PR merged first.
5. Resolves #43 - reverse URI based names for translators. Thanks everyone for the feedback in that regard.

Sorry this is such a large merge... but needed multiple rounds of feedback of the holistic design. Will try to make future PRs atomic and easier to review one at a time.

sanjaiganesh commented on org.opent2t.sample.thermostat.superpopular/org.opent2t.sample.thermostat.superpopular.xml:
```diff
@@ -0,0 +1,24 @@
+    <property name="targetTemperatureLow" type="d" access="readwrite">
+      <description>Minimum target temperature (in Celsius).</description>
+    </property>
+    
```
change units to fahrenheit

tjaffri replied:
Actually, I stuck with Celsius on purpose.... that's the SI unit. May be a little un-american of me, perhaps :)

For what it's worth, the Wink API (for which we are writing the first translator) also uses Celsius units: http://docs.winkapiv2.apiary.io/#reference/device/thermostats

#ByDesign

jasongin replied:
I agree with using Celsius units here, as long as the type is decimal not integer so there is no loss of precision when converting on either side.

Note the OCF thermostat schema that we've been discussing actually specifies the units in the property JSON.

(I am an American, and I think non-metric units are silly and old-fashioned.)

sanjaiganesh replied:
so we should fix the schemas  to "units": "C" ?

---
In reply to: [75194786](https://github.com/openT2T/translators/pull/53#discussion_r75194786) [](ancestors = 75194786)

jasongin replied:
There's not much point in making changes to the AllJoyn schemas now, until we decide whether and how we are going to align the AllJoyn schemas with OCF. Maybe we will even auto-convert them from OCF.

---
In reply to: [75208998](https://github.com/openT2T/translators/pull/53#discussion_r75208998) [](ancestors = 75208998,75194786)

tjaffri replied:
@sanjaiganesh I think this is #ByDesign. If you disagree, please file a bug.

sanjaiganesh commented:
>         "optimist": "0.6.1",

Are we not adding 'files' field ? Is it not required ?

---
Refers to: org.opent2t.sample.windowshade.superpopular/org.opent2t.test.windowshade/js/package.json:10 in 0f69203. [](commit_id = 0f692033669da96e00df195201b5162070128e1d, deletion_comment = False)

sanjaiganesh commented:
:clock1:

tjaffri commented:
@sanjaiganesh per the discussion in https://github.com/openT2T/translators/issues/51 we should put that in the build script rather than include files inside the package.json as a fixed entry. Please file a separate tracking issue if you are concerned.

jasongin commented:
>         "optimist": "0.6.1",

Any use of "files" or "main" in the package.json here is going to be wrong because of how the package.json file is in the wrong directory. Both properties will be automatically generated/overwritten by the build script anyway.

---
In reply to: [240483336](https://github.com/openT2T/translators/pull/53#issuecomment-240483336) [](ancestors = 240483336)

---
Refers to: org.OpenT2T.WindowShade.SuperPopular.Sample/org.OpenT2T.Test.WindowShade/js/package.json:10 in 0f69203. [](commit_id = 0f692033669da96e00df195201b5162070128e1d, deletion_comment = False)

jasongin commented on org.opent2t.sample.thermostat.superpopular/com.nest.thermostat/js/manifest.xml:
```diff
@@ -3,9 +3,9 @@
   <!-- associated schema -->
-  <schema id="org.OpenT2T.Thermostat.SuperPopular.Sample" />
+  <schema id="org.opent2t.sample.thermostat.superpopular" />
   <!-- associated onboarding module -->
```
Are we intending to support both a single <schema> element at the top level as well as one or more <schema> elements under a <schemas> element? The parsing code in my lib PR currently only supports the latter, but I can change it.

tjaffri replied:
This nest thermostat translator has not been updated as part of this PR (none of the other translators have been updated either). So let's ignore whatever that says (other than renaming/casing changes of course).

The Wink Thermostat is under review here. Issue #36 tracks updating all the other translators (Once we agree on this one I can do a series of PRs modifying all the other translators one by one following the same pattern)

jasongin commented:
> 

This class follows the AllJoyn schema, which is not in sync with the RAML schema of the same name. Since the schema .js file references the RAML version, this translator is not consistent with that.

---
Refers to: org.opent2t.sample.thermostat.superpopular/com.nest.thermostat/js/thingTranslator.js:91 in 0f69203. [](commit_id = 0f692033669da96e00df195201b5162070128e1d, deletion_comment = False)

jasongin commented on org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/tests/test.js:
```diff
@@ -0,0 +1,65 @@
+var delay = require('delay')
+
+// This boilerplate code is copied from transpiled typescript AVA test code, which enables usage of 'yield' keyword. 
+var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
```
I'm concerned that this, and the use of __awaiter and yield below, makes the test code look far more complicated than it needs to be. We want other developers to be able to copy this test code to use for writing tests for translators they develop. If we're not using TypeScript, then we should just write the test code in normal ES6 syntax with .then() callbacks instead of awaits.

sanjaiganesh replied:
This is .js file. I felt this was looking clean and we really want us to use a test runner like ava, for all its benefits. when we setup lab, we know how many total test cases and how many passed etc.. May be we can simplify this by removing this boilerplate code and fixing it.

---
In reply to: [75198452](https://github.com/openT2T/translators/pull/53#discussion_r75198452) [](ancestors = 75198452)

jasongin replied:
We can have all those benefits while using normal .then() callbacks instead of this __awaiter which is messy and unfamiliar to people. AVA doesn't require TypeScript or await.

---
In reply to: [75208579](https://github.com/openT2T/translators/pull/53#discussion_r75208579) [](ancestors = 75208579,75198452)

tjaffri replied:
This is a great discussion, but I'm hoping we don't need to block this PR for this. @sanjaiganesh I added issue #55 to track this... perhaps you could take a stab at it? For this PR I think this is #WontFix and we can clean it up later (very soon hopefully, perhaps tomorrow/friday).

jasongin commented on oic.baseResource/oic.baseResource.json:
```diff
@@ -0,0 +1,61 @@
+    "description": "Copyright (c) 2016 Open Connectivity Foundation, Inc. All rights reserved.",
+    "title": "Base Resource",
+    "definitions": {
+        "oic.r.baseresource": {
```
>oic.r.baseresource [](start = 9, length = 18)

There is actually a "baseResource" RAML at oneiota.org, should that be included here?

(I didn't find a "core" RAML.)

tjaffri replied:
Thanks! I'll add the RAML in a subsequent PR. Added this issue to track adding that later: https://github.com/openT2T/translators/issues/54 so for now this is #WontFix in this PR (hope I got that terminology right?)

tjaffri commented:
@jasongin as long as you see my note about the nest thermostat translator being out of scope for this PR, I think I have addressed all feedback so far.

Do you guys have more feedback?

jasongin commented:
:shipit:

sanjaiganesh commented:
:shipit:
